nexus/src/db/datastore.rs

```diff
@@ -2444,4 +2679,219 @@ mod test {
+            assert_eq!(params.extent_count(), region.extent_count());
+        }
+
+        // Allocate regions for a second disk. Observe that we allocate from
+        // the three previously unused datasets.
+        let params = create_test_disk_create_params(
+            "disk2",
+            ByteCount::from_mebibytes_u32(500),
+        );
+        let disk2_id = Uuid::new_v4();
+        let dataset_and_regions =
+            datastore.region_allocate(disk2_id, &params).await.unwrap();
+        assert_eq!(REGION_REDUNDANCY_THRESHOLD, dataset_and_regions.len());
+        let mut disk2_datasets = HashSet::new();
+        for (dataset, region) in dataset_and_regions {
+            assert!(disk2_datasets.insert(dataset.id()));
+            assert_eq!(disk2_id, region.disk_id());
+            assert_eq!(params.block_size(), region.block_size());
+            assert_eq!(params.extent_size(), region.extent_size());
+            assert_eq!(params.extent_count(), region.extent_count());
+        }
+
+        // Double-check that the datasets used for the first disk weren't
+        // used when allocating the second disk.
+        assert_eq!(0, disk1_datasets.intersection(&disk2_datasets).count());
+
+        let _ = db.cleanup().await;
+    }
+
+    #[tokio::test]
+    async fn test_region_allocation_is_idempotent() {
+        let logctx =
+            dev::test_setup_log("test_region_allocation_is_idempotent");
+        let mut db = test_setup_database(&logctx.log).await;
+        let cfg = db::Config { url: db.pg_config().clone() };
+        let pool = db::Pool::new(&cfg);
+        let datastore = DataStore::new(Arc::new(pool));
+
+        // Create a sled...
+        let sled_id = create_test_sled(&datastore).await;
+
+        // ... and a zpool within that sled...
+        let zpool_id = create_test_zpool(&datastore, sled_id).await;
+
+        // ... and datasets within that zpool.
+        let dataset_count = REGION_REDUNDANCY_THRESHOLD;
+        let bogus_addr =
+            SocketAddr::new(IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)), 8080);
+        let kind =
+            DatasetKind(crate::internal_api::params::DatasetKind::Crucible);
+        let dataset_ids: Vec<Uuid> =
+            (0..dataset_count).map(|_| Uuid::new_v4()).collect();
+        for id in &dataset_ids {
+            let dataset = Dataset::new(*id, zpool_id, bogus_addr, kind.clone());
+            datastore.dataset_upsert(dataset).await.unwrap();
+        }
+
+        // Allocate regions from the datasets for this disk.
+        let params = create_test_disk_create_params(
+            "disk",
+            ByteCount::from_mebibytes_u32(500),
+        );
+        let disk_id = Uuid::new_v4();
+        let mut dataset_and_regions1 =
+            datastore.region_allocate(disk_id, &params).await.unwrap();
+        let mut dataset_and_regions2 =
+            datastore.region_allocate(disk_id, &params).await.unwrap();
```

Comment:
When I first read this, I thought we were testing that two different allocations would pick the same regions (which seems undesirable).  I gather what you're really testing is that if we attempt to allocate again for the same disk, we get the same regions.  Is it worth adding a separate attempt to allocate regions for a second disk, between the two allocations for disk1, to ensure that we're normally _not_ picking the same thing, but we're clearly returning what we previously picked when we make the second attempt to allocate for disk1?